files-csharp/ComparingTwoFiles/ComparingTwoFiles/FileComparer.cs

```diff
@@ -0,0 +1,56 @@
+namespace ComparingTwoFiles;
+
+public static class HashStore
+{
+    public static Dictionary<string, string> Hashes { get; } = new()
+    {
+        ["files/batch1/hello-world.txt"] = "A0752635617DC83B5858BEDC41E22F91",
+        ["files/batch2/hello-world.txt"] = "4FCEE0F97E586FD2D9111397BC9D0440"
+    };
+}
+
+public class FileComparer
+{
+    public static bool CompareByNameAndSize(string file1, string file2)
+    {
+        var fileInfo1 = new FileInfo(file1);
+        var fileInfo2 = new FileInfo(file2);
+
+        return fileInfo1.Name == fileInfo2.Name && fileInfo1.Length == fileInfo2.Length;
+    }
+
+    public static bool CompareByBytes(string file1, string file2)
+    {
+        var bytes1 = File.ReadAllBytes(file1);
+        var bytes2 = File.ReadAllBytes(file2);
+
+        if (bytes1.Length != bytes2.Length)
+        {
+            return false;
+        }
+
+        return Enumerable.SequenceEqual(bytes1, bytes2);
+    }
+
+    public static bool CompareByChecksum(string file1, string file2)
+    {
+        var hash1 = MD5.HashData(File.ReadAllBytes(file1));
+        var hash2 = MD5.HashData(File.ReadAllBytes(file2));
+
+        return Enumerable.SequenceEqual(hash1, hash2);
+    }
```

Comment:
Rather then reading the entire contents of the file into an array and then processing, both of these methods could employ FileStream.

In the checksum method, we can just pass the stream directly into the MD5 hash method. Since we are looking at performance in this code (benchmarking), we could stackalloc the resultant array, thus we would remove all heap allocation in the checksum method.

In the CompareByBytes method, again, we could use FileStreams, checking length first, then allocate two buffers for reading, and then read the files in chunks, comparing each chunk. This way if the files differ early on, we bail before reading the entire file. This also has the advantage of minimizing allocation. For instance if the files being compared where 100 MB in length, instead of reading 200 MB of data into arrays and then comparing the arrays, we could open the streams and read 1K chunks or something like that into Span buffers, and then compare those chunks using SequenceEquals.